Mathlib/Topology/Bornology/Absorbs.lean

```diff
@@ -0,0 +1,245 @@
+/-
+Copyright (c) 2023 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Data.Set.Pointwise.SMul
+import Mathlib.Topology.Bornology.Constructions
+
+#align_import analysis.locally_convex.basic from "leanprover-community/mathlib"@"f2ce6086713c78a7f880485f7917ea547a215982"
+
+/-!
+# Definitions of `Absorbs` and `Absorbent`
+
+Let `M` act on `α`, let `A` and `B` be sets in `α`.
+We say that `A` *absorbs* `B` if for sufficiently large `a : M`, we have `B ⊆ a • A`.
+Formally, "for sufficiently large `a : M`" means "for all but a bounded set of `a`".
+
+Traditionally, this definition is formulated
+for the action of a (semi)normed ring on a module over that ring.
+
+We formulate it in a more general settings for two reasons:
+
+- this way we don't have to depend on metric spaces, normed rings etc;
+- some proofs look nicer with this definition than with something like
+  `∃ r : ℝ, ∀ a : R, r ≤ ‖a‖ → B ⊆ a • A`.
+
+-/
+
+open Set Bornology Filter
+open scoped Pointwise
+
+assert_not_exists Real
+
+section Defs
+
+variable (M : Type*) {α : Type*} [Bornology M] [SMul M α]
+
+/-- A set `s` absorbs another set `t` if `t` is contained in all scalings of `s`
+by all but a bounded set of elements. -/
+def Absorbs (s t : Set α) : Prop :=
+  ∀ᶠ a in cobounded M, t ⊆ a • s
+#align absorbs Absorbs
+
+/-- A set is *absorbent* if it absorbs every singleton. -/
+def Absorbent (s : Set α) : Prop :=
+  ∀ x : α, Absorbs M s {x}
+#align absorbent Absorbent
+
+end Defs
+
+namespace Absorbs
+
+section SMul
+
+variable {M α : Type*} [Bornology M] [SMul M α] {s t : Set α}
+
+protected lemma empty : Absorbs M s ∅ := by simp [Absorbs]
+#align absorbs_empty Absorbs.empty
+
+protected lemma eventually (h : Absorbs M s t) : ∀ᶠ a in cobounded M, t ⊆ a • s := h
+
+@[simp] lemma of_boundedSpace [BoundedSpace M] : Absorbs M s t := by simp [Absorbs]
+
+lemma mono_left {s'} (h : Absorbs M s t) (hs : s ⊆ s') : Absorbs M s' t :=
+  h.mono fun _a ha ↦ ha.trans <| smul_set_mono hs
+#align absorbs.mono_left Absorbs.mono_left
+
+lemma mono_right {t'} (h : Absorbs M s t) (ht : t' ⊆ t) : Absorbs M s t' :=
+  h.mono fun _ ↦ ht.trans
+#align absorbs.mono_right Absorbs.mono_right
+
+lemma mono {s' t'} (h : Absorbs M s t) (hs : s ⊆ s') (ht : t' ⊆ t) : Absorbs M s' t' :=
+  (h.mono_left hs).mono_right ht
+#align absorbs.mono Absorbs.mono
+
+@[simp]
+lemma _root_.absorbs_union {t₁ t₂} : Absorbs M s (t₁ ∪ t₂) ↔ Absorbs M s t₁ ∧ Absorbs M s t₂ := by
+  simp [Absorbs]
+#align absorbs_union absorbs_union
```

Comment:
Could you extract more variables to the `variable` statement and make up your mind between using `t`, `t'` vs `t₁`, `t₂` (if you ask me, I prefer the latter)?

Reply:
AFAIR, I copied lemmas from different places. Thank you, fixed!